jena-text/src/main/java/org/apache/jena/query/text/TextIndexES.java

```diff
@@ -0,0 +1,394 @@
+        Map<String, Node> result = new HashMap<>();
+
+        if(uri != null) {
+            response = client.prepareGet(indexName, docDef.getEntityField(), uri).get();
+            if(response != null && !response.isSourceEmpty()) {
+                String entityField = response.getId();
+                Node entity = NodeFactory.createURI(entityField) ;
+                result.put(docDef.getEntityField(), entity);
+                Map<String, Object> source = response.getSource();
+                for (String field: docDef.fields()) {
+                    Object fieldResponse = source.get(field);
+
+                    if(fieldResponse == null) {
+                        //We wont return it.
+                        continue;
+                    }
+                    else if(fieldResponse instanceof List<?>) {
+                        //We are storing the values of fields as a List always.
+                        //If there are values stored in the list, then we return the first value,
+                        // else we do not include the field in the returned Map of Field -> Node Mapping
+                        List<?> responseList = (List<?>)fieldResponse;
+                        if(responseList != null && responseList.size() > 0) {
+                            String fieldValue = (String)responseList.get(0);
+                            Node fieldNode = NodeFactoryExtra.createLiteralNode(fieldValue, null, null);
+                            result.put(field, fieldNode);
+                        }
+                    }
+                }
+            }
+        }
+
+        return result;
+    }
+
+    @Override
+    public List<TextHit> query(Node property, String qs) {
+
+        return query(property, qs, MAX_RESULTS);
+    }
+
+    /**
+     * Query the ElasticSearch for the given Node, with the given query String and limit.
+     * @param property the node property to make a search for
+     * @param qs the query string
+     * @param limit limit on the number of records to return
+     * @return List of {@link TextHit}s containing the documents that have been found
+     */
+    @Override
+    public List<TextHit> query(Node property, String qs, int limit) {
+
+        qs = parse(qs);
+        LOGGER.debug("Querying ElasticSearch for QueryString: " + qs);
+        SearchResponse response = client.prepareSearch(indexName)
+                .setTypes(docDef.getEntityField())
+                .setQuery(QueryBuilders.queryStringQuery(qs))
+                .setFrom(0).setSize(limit)
+                .get();
+
+        List<TextHit> results = new ArrayList<>() ;
+        for (SearchHit hit : response.getHits()) {
+
+            Node literal;
+            String field = (property != null) ? docDef.getField(property) : docDef.getPrimaryField();
+            List<String> value = (List<String>)hit.getSource().get(field);
+            if(value != null) {
+                literal = NodeFactory.createLiteral(value.get(0));
```

Comment:
So Just to be on the same page, here is my current Add and Query Algorithm. Can you please clarify my assumptions.

_ASSUMPTION_: `rdfs:label` is mapped to a field with name: `label` in my Index
## _*ADD Functionality*_
* If I get a triple like this: `:x1 rdfs:label "X2 word"@en .` I store it in ES under the field: `label_en` ONLY with value : `["X2 word"]` . Note that it is an Array.
* Next, If I get the next triple as : `:x1 rdfs:label "X1 another word" .`  , I store it in ES under the field : `label` ONLY with value: `["X1 another word"]` . Note that it is an Array.
Thus at this point my ES Document with id: `http://example/x1` has two fields: `label` and `label_en`.
* Next I get another triple as: `:x1 rdfs:label "X3 yet another word"@en .`, I store it in ES under the field: `label_en` ONLY. Now the `label_en` field has values: `["X2 word", "X3 yet another word"]` 
* Finally I get another triple as: `:x1 rdfs:label "X3 Simple" .`, I store it in ES under the field: `label` ONLY. Now the `label` field has values: `["X1 another word", "X3 Simple"]` 

Let's assume the above is the state of the Index at Query Time. i.e. I have ONE Document in ES with id: `http://example/x1` and that this document has two fields:
`label`:`["X1 another word", "X3 Simple"]`
`label_en`:`["X2 word", "X3 yet another word"]`

## _QUERY Functionality_
### CASE 1: When Query is `('X3')`
In this case the ES implementation will see that there is no language specific search required by the user and thus it will create a text query that will search in all the related fields, i.e language-specific AS WELL AS language-agnostic. The text query will have the syntax: `X1`
Thus the results returned will contain both `label` as well as `label_en` field.
Next since the Node Property is NULL for the above query, I will take the `docDef.getPrimaryField()`  (which in may case has a value `label`) as the field whose value I should return. Thus I take the value of `label` field which is `["X1 another word", "X3 Simple"]` and currently return the first value: `X1 another word`. Clearly this is NOT the right Value that I should return. I should instead return ALL the literals that have `X3` as a string in them. Is that the correct assumption?

###Case 2: When Query is `rdfs:label 'X3'`
As above I am doing both language-specific as well as language-agnostic query and the actual query string will look like: `label*:X3`. Thus as above I will return the wrong value `X1 another word`.  I should instead return ALL the literals that have `X3` as a string in them. Is that the correct assumption?

###Case 3: When Query is `rdfs:label 'X3' lang:en`
In this case, the query String will look like: `label_en:X3`. Thus I am searching for ONLY language specific field and NOT all the fields. Once I get the result, I then return the first value ONLY. I should instead return ALL the literals that have `X3` as a string in them. Is that the correct assumption?